addons/project_sale_expense/models/hr_expense.py

```diff
@@ -1,16 +1,13 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models
+from odoo import api, models
 
 
 class Expense(models.Model):
     _inherit = "hr.expense"
 
+    @api.depends('sale_order_id')
     def _compute_analytic_distribution(self):
         super()._compute_analytic_distribution()
         for expense in self:
-            if not expense.sale_order_id:
-                continue
-            analytic_distribution = expense.sale_order_id.project_id._get_analytic_distribution()
-            if analytic_distribution:
-                expense.analytic_distribution = analytic_distribution
+            expense.analytic_distribution = expense.sale_order_id.project_id._get_analytic_distribution()
```

Comment:
I couldn't get `sale_order_id` to be `False` but an empty recordset instead :shrug: So the problem does not arise because `_get_analytic_distribution()` returns an empty dict in that case. And if the sale order has no project, then there is no problem too (it also returns an empty dict). 

I have added some tests for that just in case :)